### PR TITLE
Pool query objects

### PR DESCRIPTION
**Pull request by i**
Queries account for a lot of unnecessary allocations. This PR exposes a `Release` method on the `Query` type so that callers of `Session.Query()` can reduce the number of allocations.

### CODE REVIEW

**Comment by nemothekid** [comment]
On a very quick look through, doesn't it make more sense to eschew the `Release` method and just invalidate/`Put` the `Query` reference after `Close()` is called?

File: session.go [hunk 1]
```diff
@@ -98,11 +100,12 @@ func NewSession(cfg ClusterConfig) (*Session, error) {
...
-		prefetch: 0.25,
-		cfg:      cfg,
-		pageSize: cfg.PageSize,
-		stmtsLRU: &preparedLRU{lru: lru.New(cfg.MaxPreparedStmts)},
+		cons:      cfg.Consistency,
+		prefetch:  0.25,
+		cfg:       cfg,
+		pageSize:  cfg.PageSize,
+		stmtsLRU:  &preparedLRU{lru: lru.New(cfg.MaxPreparedStmts)},
+		queryPool: &sync.Pool{New: func() interface{} { return new(Query) }},
```
Comment (Zariel) [inline session.go]: this pool can be package local

File: session.go [hunk 2]
```diff
@@ -947,6 +956,44 @@ func (q *Query) MapScanCAS(dest map[string]interface{}) (applied bool, err error
...
+		return
+	}
+	pool := q.session.queryPool
+	q.reset()
+	pool.Put(q)
+}
+
+// reset zeroes out all fields of a query so that it can be safely pooled.
+func (q *Query) reset() {
+	q.stmt = ""
```
Comment (Zariel) [inline session.go]: i dont think all these fields need to be reset, only ones which dont get set by `Query()`

**Comment by Zariel** [comment]
I agree with @nemothekid it would be better to do this in Close and update the documentation to state that the Query must not be reused after close is called.

**Comment by i** [comment]
@nemothekid and @Zariel,

Are you guys referring to to `Iter.Close()`? What happens in the case where a query is used more than once? Although it isn't explicitly supported, I don't think it's unreasonable to do something like this in order to reuse a query:
```go
qry := session.Query("select * from health_check")
for ;; time.Sleep(time.Second) {
  if err := qry.Exec(); err != nil {
    ...
  }
}

```

Am I wrong in my assumption?

File: session.go [hunk 3]
```diff
@@ -947,6 +956,44 @@ func (q *Query) MapScanCAS(dest map[string]interface{}) (applied bool, err error
...
 
+// Release releases a query back into a pool of queries. Released Queries
+// cannot be reused.
+//
+// Example:
+// 		qry := session.Query("SELECT * FROM my_table")
+// 		qry.Exec()
+// 		qry.Release()
+func (q *Query) Release() {
+	if q == nil || q.session == nil || q.session.queryPool == nil {
```
Comment (Zariel) [inline session.go]: this can be simplified if the pool is a package local variable then it will never be nil, how worthwhile is the query nil check?

**Comment by Zariel** [comment]
Ah sorry I miss understood/forgot how the query api was working in the case without iterators:)

Patch looks good once the suggested changes to move the pool out of session. If you are not already could you please add yourself to the AUTHORS file and ill merge once the changes have been made.

File: session.go (hunk 2 above)
Comment (i) [inline session.go]: Assuming this is invalid since we're pooling at the package-level now.

File: session.go [hunk 4]
```diff
@@ -947,6 +959,40 @@ func (q *Query) MapScanCAS(dest map[string]interface{}) (applied bool, err error
...
+// Release releases a query back into a pool of queries. Released Queries
+// cannot be reused.
+//
+// Example:
+// 		qry := session.Query("SELECT * FROM my_table")
+// 		qry.Exec()
+// 		qry.Release()
+func (q *Query) Release() {
+	q.reset()
+	pool.Put(q)
```
Comment (Zariel) [inline session.go]: this should be `queryPool.Put(q)`

Comment (i) [inline session.go]: good catch, thanks

**Comment by Zariel** [comment]
Thanks very much for the contribution :)

**Comment by i** [comment]
Thanks for the opportunity 😃

**Pull request merged**